src/harness/testcases/WINNF_FT_S_FAD_testcase.py

```diff
@@ -285,27 +289,61 @@ def test_WINNF_FT_S_FAD_1(self, config_filename):
       # Register N1 CBSDs.
       request = {'registrationRequest': config['registrationRequests']}
       responses = self._sas.Registration(request)['registrationResponse']
+
       # Check registration responses and get cbsd Id
       self.assertEqual(len(responses), len(config['registrationRequests']))
-      # TODO : to update the test so that it is possible to get a PAL Grant
+      cbsd_ids = []
+      for response in responses:
+        cbsd_ids.append(response['cbsdId'])
+      cbsdIds_of_pal_grant = []
+      # inject PALs and N2 PPAs
+      ppa_ids = []       
+      for pal in config['palRecords']:
+          self._sas_admin.InjectPalDatabaseRecord(pal)                  
+      for ppa in config['ppaRecords']:
+        ppa['ppaInfo']['cbsdReferenceId'] = []
+        for cluster_list in ppa['cbsdClusterList']:
+          for index, cbsd in enumerate(config['registrationRequests']):
+            if cluster_list['fccId'] == cbsd['fccId'] and  cluster_list['cbsdSerialNumber'] == cbsd ['cbsdSerialNumber']:
+              ppa['ppaInfo']['cbsdReferenceId'].append(cbsd_ids[index])
+              break
+        del ppa['cbsdClusterList']
+        ppa_ids.append(self._sas_admin.InjectZoneData({'record': ppa}))
       grants = config['grantRequests']
       for index, response in enumerate(responses):
           self.assertEqual(response['response']['responseCode'], 0)
           grants[index]['cbsdId'] = response['cbsdId']
+          grant_frequency_range = grants[index]['operationParam']['operationFrequencyRange']
+          # Get the cbsdIds whose grant is PAL
```

Comment:
Can't you just compare the `channelType` in the `grantResponse` with the `channelType` in the FAD? We have other tests to check that a PAL grant is issued correctly, so we don't need to double-check here.

Reply:
it is a good remark!
I kept the check for consistent PPA&PAL in the config file at the begning of the test, if you think it is not necessary, I can delete it.

Reply:
Thanks!

I like the idea of checking the config more carefully, but unfortunately I'm not sure that the checks are valid. For example, we could configure the test case to give a GAA Grant to a device which is on the PPA cluster list, and that would be a completely valid config (and in fact it might be a good test of the SAS).

I'm not sure what checks you *can* do, other than some obvious ones:
 - Grants are for valid channels
 - CBSD X doesn't belong to two different PPA cluster lists
 - Each PPA has a cluster list (?)

Reply:
the check here is only to verify the following:
- the PAL channels of one PPA form  a frequencyRange correctly.
- if the CBSD is in the cluster list of a PPA then the grant  frequency range should be completely inside the frequency range of the PPA or completely outside of it, so it should not partially overlap with it.

don't you agree with that?